crates/worker/src/partition/state_machine/command_interpreter.rs

```diff
@@ -805,38 +806,40 @@ where
                 }
             }
             EnrichedEntryHeader::CompleteAwakeable {
-                ref invocation_id,
-                entry_index,
+                enrichment_result:
+                    AwakeableEnrichmentResult {
+                        invocation_id,
+                        entry_index,
+                    },
+                ..
             } => {
-                let_assert!(Entry::CompleteAwakeable(entry) = Codec::deserialize(&journal_entry)?);
+                let_assert!(Entry::CompleteAwakeable(entry) = journal_entry.as_entry::<Codec>()?);
 
                 self.send_message(
                     OutboxMessage::from_awakeable_completion(
                         invocation_id.clone(),
-                        entry_index,
+                        *entry_index,
                         entry.result.into(),
                     ),
                     effects,
                 );
             }
-            EnrichedEntryHeader::Custom { requires_ack, code } => {
-                if requires_ack {
-                    // Reset the requires_ack flag to false
-                    journal_entry.header = EnrichedEntryHeader::Custom {
-                        code,
-                        requires_ack: false,
-                    };
-                    effects.forward_completion(
-                        full_invocation_id.clone(),
-                        Completion {
-                            entry_index,
-                            result: CompletionResult::Ack,
-                        },
-                    );
-                }
+            EnrichedEntryHeader::Custom { .. } => {
+                // We just store it
             }
         }
 
+        if journal_entry.header().requires_ack() {
+            // Reset the requires_ack flag to false
+            journal_entry.header_mut().mark_acked();
+            effects.forward_completion(
+                full_invocation_id.clone(),
+                Completion {
+                    entry_index,
+                    result: CompletionResult::Ack,
+                },
+            );
```

Comment:
The storage ack is not the same thing, that's the signal for the invoker. You mean I should create a new effect?

One thing I would like to do though is to rename `Completion` and `CompletionResult` and all the related names to "notification"/"journal notification". notifications include acks and completions.

Reply:
I guess what I mean is to not encode `ProtocolMessage::EntryAck` via a `CompletionResult::Ack`. I think this is a bit misleading because technically it is not a completion. Renaming the whole mechanims into `Notifications` might already do the trick.

One observation is that we are already acking the storage to the invoker via the storage ack. Why do we actually need a separate signal to ack the entry to the service deployment? Couldn't this be the very same signal? In that case, we wouldn't even have to surface the `requires_ack` field to the partition processor and could stay an invoker concept only.

Reply:
I think you're right on that, perhaps we should keep everything related to this entry ack something "invoker internal". IIRC one of the reasons we didn't do it like that at the beginning is that we considered very different the invoker storage entry ack (and potentially we wanted to get rid of it somehow?!).

But happy to revert that change, and move all the requires_ack thing inside the invoker.